### PR TITLE
Test Trino S3 filesystem with MinIO

### PR DESCRIPTION
**Pull request by findepi**
MinIO is important deployment context, so let's test S3 filesystem with MinIO as well.

### CODE REVIEW

File: lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/TestS3FileSystemMinIo.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+                .hasMessage("S3 HEAD request failed for file: s3://" + bucket + "/test/.././/file");
+    }
+
+    @Test
+    @Override
+    public void testListFiles()
+            throws IOException
+    {
+        // MinIO is not hierarchical but has hierarchical naming constraints. For example it's not possible to have two blobs "level0" and "level0/level1".
+        testListFiles(true, true);
```
Comment (losipiuk) [inline lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/TestS3FileSystemMinIo.java:107]: OH - so it should be `false,true`?

File: lib/trino-filesystem/src/test/java/io/trino/filesystem/AbstractTestTrinoFileSystem.java [hunk 2]
```diff
@@ -774,11 +781,17 @@ void testRenameFile()
...
 
     @Test
-    void testListFiles()
+    public void testListFiles()
+            throws IOException
+    {
+        testListFiles(isHierarchical(), isHierarchical());
+    }
+
+    protected void testListFiles(boolean hierarchicalNamingConstraints, boolean hierarchicalDirectoryNaming)
```
[the remark this answers is not in the document]
Comment (findepi) [reply lib/trino-filesystem/src/test/java/io/trino/filesystem/AbstractTestTrinoFileSystem.java]: replaced with just `hierarchicalNamingConstraints`

File: lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/TestS3FileSystemMinIo.java (hunk 1 above)
Comment (findepi) [reply lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/TestS3FileSystemMinIo.java]: not exactly, 
the first boolean isn't "is hierachical" (which MinIO isn't)
but it's "has hierarchical naming constraints" (which MinIO is)

**Comment by findepi** [comment]
Thanks for review, AC. PTAL!

**Review by losipiuk — APPROVED** [review]

**Review by ebyhr — APPROVED** [review]
Looks good except for CI failure:
```
/home/runner/work/trino/trino/lib/trino-filesystem-azure/src/test/java/io/trino/filesystem/azure/AbstractTestAzureFileSystem.java:[174,10] testPaths() in io.trino.filesystem.azure.AbstractTestAzureFileSystem cannot override testPaths() in io.trino.filesystem.AbstractTestTrinoFileSystem
```

File: lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/TestS3FileSystemMinIo.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    {
+        return bucket;
+    }
+
+    @Override
+    protected S3Client createS3Client()
+    {
+        try {
+            return S3Client.builder()
+                    .endpointOverride(new URI(minio.getMinioAddress()))
```
Comment (electrum) [inline lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/TestS3FileSystemMinIo.java:67]: Use `URI.create()` which throws an unchecked exception

File: lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/TestS3FileSystemMinIo.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import static io.trino.testing.TestingNames.randomNameSuffix;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class TestS3FileSystemMinIo
+        extends AbstractTestS3FileSystem
+{
+    private final String bucket = "test-bucket-" + randomNameSuffix();
```
Comment (electrum) [inline lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/TestS3FileSystemMinIo.java:35]: Use a fixed bucket name to be consistent with the other tests

**Comment by findepi** [comment]
Just found this class... Does this obsolete `io.trino.plugin.hive.s3.TestTrinoS3FileSystemMinio` @findinpath ?

**Comment by findinpath** [comment]
> Does this obsolete io.trino.plugin.hive.s3.TestTrinoS3FileSystemMinio @findinpath ?

@findepi Not yet. We would have to move the `20` tests from `io.trino.plugin.hive.s3.BaseTestTrinoS3FileSystemObjectStorage` towards `io.trino.filesystem.s3.AbstractTestS3FileSystem`  / `AbstractTestTrinoFileSystem` to make the previously mentioned class obsolete.

**Pull request merged**